src/main/java/yesman/epicfight/world/capabilities/entitypatch/HurtableEntityPatch.java

```
@@ -95,8 +95,13 @@ public void setDefaultStunReduction(EquipmentSlot equipmentslot, ItemStack from,
 		List<AttributeModifier> modifiersToAdd = CapabilityItem.getAttributeModifiersAsWeapon(EpicFightAttributes.STUN_ARMOR, equipmentslot, to, null);
 		List<AttributeModifier> modifiersToRemove = CapabilityItem.getAttributeModifiersAsWeapon(EpicFightAttributes.STUN_ARMOR, equipmentslot, from, null);
 		
+		AttributeInstance stunArmorAttribute = this.original.getAttribute(EpicFightAttributes.STUN_ARMOR);
+		if (stunArmorAttribute == null) {
+			return;
+		}
```

Is there a chance that this could make the entity misbehave instead of crashing the game?

Sometimes, it's better to crash than return without any sign of issues, so I'm wondering if this is the case here.

---

Not all entities are intended to have the attribute in the first place, that is why it is null sometimes. You [collective] do not register it globally for all `LivingEntity` objects, right? I think a better way to think about it from here is, if the entity lacks `STUN_ARMOR` attribute, it is by design not compatible with the mechanic. In that case, entities will effectively have 0 stun reduction, which is fine. This is actually kinda how vanilla behaves with respect to attributes: if it is not present, there is no container for modifiers.

To be honest, I completely disagree that logic of failing fast as a good outcome applies to games in general; games are special in that regard its much more annoying to crash the game than to have a behavioural bug which nonetheless is not a showstopper. Unfortunately, the code has little defensive programming measures, no behavioural tests... In this case, the worst that will happen is that some bugged entities will not have stun reduction; the default of 0 will be used. At least, this keeps the game playable.

---

> I completely disagree that logic of failing fast as a good outcome applies to games in general; 

In some cases, an alternative is to log an error or debug, which can be used to debug issues later (example #1935), instead of silently misbehaving.

Imagine a system where everything stops working after years of work, and with no sign of errors, that would be pretty hard to track and debug.
However, this is not the case here.

Thank you for the detailed answer.

---

I think we should deeply inspect what causes the crash exactly. Logically, this shouldn't be null since I force injected Epic Fight Attributes in [MixinLivingEntity](https://github.com/Epic-Fight/epicfight/blob/1.21.1/src/main/java/yesman/epicfight/mixin/common/MixinLivingEntity.java#L56C3-L56C41)

The code indicates all entities with `HurtableEntityPatch` will have `EpicFightAttributes.STUN_ARMOR` attribute instance. You can find the fact in vanilla `AttributeMap` class.

```java
@Nullable
    public AttributeInstance getInstance(Holder<Attribute> attribute) {
        return (AttributeInstance)this.attributes.computeIfAbsent(attribute, (p_315936_) -> this.supplier.createInstance(this::onAttributeModified, p_315936_));
    }
```

The `supplier` object is where I replaced at the mixin code

```java
    self.getAttributes().supplier = builder.build();
````

So, if `this.original.getAttribute(EpicFightAttributes.STUN_ARMOR)` returns null, it means my mixin didn't work at all or the attribute supplier is replaced by another mixin from other mods.

@0x484558  Would you present the case where you had trouble?

---

The existence of the null check in getStunArmor() suggested to me that it is known that the attribute might be missing despite the mixin.

---

FYI: I have added a [new guideline](https://github.com/Epic-Fight/epicfight/commit/20f2c02ea7e5c367cb5ac8f5338249baba8fe324) to document this guarding pattern.